battle-service/src/main/java/org/astashonok/battleservice/services/impl/BattleServiceImpl.java

```diff
@@ -0,0 +1,245 @@
+    @Autowired
+    private MoveRepository moveRepository;
+
+    @Autowired
+    private BattleCreationFormValidator battleCreationFormValidator;
+
+    @Override
+    public void makeMove(MoveForm moveForm) {
+        asserts(() -> moveFormValidator.validate(moveForm), MoveException::new);
+
+        Optional.of(moveForm.getBattleId()).flatMap(battleId -> battleRepository.findById(battleId)).map(battle -> {
+            asserts(() -> battleValidator.validate(battle), BattleException::new);
+            return battle;
+        }).ifPresent(battle -> makeMove(battle, moveForm));
+    }
+
+    @Override
+    public List<Battle> getOpenedBattles() {
+        return battleRepository.getAllByStatus(BattleStatus.OPENED);
+    }
+
+    @Override
+    public void join(Long battleId) {
+        Optional.ofNullable(battleId).flatMap(battleRepository::findById).map(battle -> {
+            battle.setOParticipantId(authUserService.getCurrentUserId());
+            battle.setStatus(BattleStatus.IN_PROCESS);
+            return battle;
+        }).map(battleRepository::save).orElseThrow(() -> new BattleException(String.format(BATTLE_NOT_EXISTS, battleId)));
+    }
+
+    @Override
+    public Battle openBattle(BattleCreationForm form) {
+        return battleRepository.save(battleFactory.create(form));
+    }
+
+    @Override
+    public BattleInfo getBattleInfo(Long battleId) {
+        return Optional.ofNullable(battleId)
+                .flatMap(battleRepository::findById)
+                .map(battle -> new BattleInfo(battle.getStatus(), battle.getWinnerId()))
+                .orElseThrow(() -> new BattleException(String.format(BATTLE_NOT_EXISTS, battleId)));
+    }
+
+    private void asserts(Supplier<List<String>> validate, Function<String, RuntimeException> exceptionCreator) {
+        List<String> messages = validate.get();
+        if (!CollectionUtils.isEmpty(messages)) {
+            throw exceptionCreator.apply(messages.toString());
+        }
+    }
+
+    private void makeMove(Battle battle, MoveForm moveForm) {
+        battle.setFieldNumberValue(authUserService.getCurrentUserId(), moveForm.getXCoordinate(), moveForm.getYCoordinate());
+        battle.decrementRemainingFreeMoveCount();
+        battle.swapNextMoveParticipantId();
+        battle.setWinnerId(getWinnerId(battle));
+        if (battle.getWinnerId() != null || battle.getRemainingFreeMoveCount() == 0) {
+            battle.setStatus(BattleStatus.FINISHED);
+        }
+        battleRepository.save(battle);
+        saveUserMove(moveForm);
+    }
+
+    private void saveUserMove(MoveForm moveForm) {
+        Move move = new Move();
+        move.setBattleId(moveForm.getBattleId());
+        move.setUserId(authUserService.getCurrentUserId());
+        move.setXCoordinate(moveForm.getXCoordinate());
+        move.setYCoordinate(moveForm.getYCoordinate());
+        move.setTime(new Date());
+
+        moveRepository.save(move);
+    }
+
+    private Long getWinnerId(Battle battle) {
```

Comment:
something with naming? 
why winner get calculated each time when you request this method?

Reply:
It's because when player makes move the game can be finished. And to determine winner (if he is) and to set required battle status I call this method inside a 'makeMove()' method